main/src/ca/uwaterloo/flix/language/phase/unification/FastBoolUnification.scala

```diff
@@ -0,0 +1,1715 @@
+    */
+  private def propagateAnd(t: Term): Term = {
+    // Simplifies the given term `t` assuming that `trueCsts` and `trueVars` are all TRUE.
+    def visit(t: Term, trueCsts: SortedSet[Int], trueVars: SortedSet[Int]): Term = t match {
+      case Term.True => Term.True
+      case Term.False => Term.False
+      case Term.Cst(c) => if (trueCsts.contains(c)) Term.True else Term.Cst(c) // `c` holds so we can replace it by TRUE.
+      case Term.Var(x) => if (trueVars.contains(x)) Term.True else Term.Var(x) // `x` holds so we can replace it by TRUE.
+      case Term.Not(t0) => Term.mkNot(visit(t0, trueCsts, trueVars))
+      case Term.And(csts0, vars0, rest0) =>
+        // Compute the constants and variables that _must_ hold for the whole conjunction to hold.
+        val termCsts = csts0.map(_.c)
+        val termVars = vars0.map(_.x)
+
+        // Extend trueCsts and trueVars.
+        val currentCsts = trueCsts ++ termCsts
+        val currentVars = trueVars ++ termVars
+
+        // Recurse on the sub-terms with the updated maps.
+        val rest = rest0.collect {
+          case t: Term.Not => visit(t, currentCsts, currentVars)
+          case t: Term.And => visit(t, currentCsts, currentVars)
+          case t: Term.Or => visit(t, currentCsts, currentVars)
+          case _: Term.Cst =>
+            // Cannot happen because the invariant of [[Term.mkAnd]] ensures there are no constants in `rest`.
+            throw InternalCompilerException("Unexpected constant", SourceLocation.Unknown)
+          case _: Term.Var =>
+            // Cannot happen because the invariant of [[Term.mkAnd]] ensures there are no variables in `rest`.
+            throw InternalCompilerException("Unexpected variable", SourceLocation.Unknown)
+        }
+
+        // Compute new constant and variable sets by removing constants and variables that hold.
+        val csts = termCsts -- trueCsts
+        val vars = termVars -- trueVars
+
+        // Recompose the conjunction. We use the smart constructor because some sets may have become empty.
+        Term.mkAnd(csts.toList.map(Term.Cst) ++ vars.toList.map(Term.Var) ++ rest)
+
+      case Term.Or(ts) => Term.mkOr(ts.map(visit(_, trueCsts, trueVars)))
+    }
+
+    visit(t, SortedSet.empty, SortedSet.empty)
+  }
+
+  /**
+    * Companion object for [[Equation]].
+    */
+  object Equation {
+    /**
+      * Returns a unification equation  `t1 ~ t2` between the terms `t1` and `t2`.
+      *
+      * The smart constructor performs normalization:
+      * - We move true and false to the rhs.
+      * - We move a variable to the lhs (unconditionally).
+      * - We reorder constant/variables equations so that the smaller constant/variable is on the lhs.
+      *
+      * Examples:
+      * -     true ~ x7 ==> x7 ~ true
+      * -       c3 ~ c2 ==> c2 ~ c3
+      * -       x7 ~ x5 ==> x5 ~ x7
+      * - x3 /\ x7 ~ x4 ==> x4 ~ x3 /\ x7
+      * - x1 /\ x2 ~ c5 ==> c5 ~ x1 /\ x2
```

Comment:
This example doesnt happen according to the code

Reply:
Fixed